### ISSUE TITLE
undefined function ExAws.Dynamo

### ISSUE THREAD

**Issue by schummjm** [CargoSense/ex_aws#215]
Newbie here - trying to get this up and running with my Phoenix app.

I keep messing with settings to try to get this working in the console, but keep getting this error:

```
iex(1)> alias ExAws.Dynamo  
nil
iex(2)> Dynamo.create_table("Users", "email", %{email: :string}, 1, 1) |> ExAws.request!
** (UndefinedFunctionError) undefined function ExAws.Dynamo.create_table/5 (module ExAws.Dynamo is not available)
    ExAws.Dynamo.create_table("Users", "email", %{email: :string}, 1, 1)
iex(2)> 
```

It seems to me it isn't finding the library at all.

Can someone help point me in the right direction to solving this error?

**Comment by benwilson512** [comment]
Hey there! Can you gist your `mix.exs` and `mix.lock` files?

**Comment by schummjm** [comment]
Hey! Thanks for taking a look.  Let me know if you need more info.

mix.exs
`  def application do
    [mod: {Sage, []},
     applications: [:phoenix, :phoenix_pubsub, :phoenix_html, :cowboy, :logger, :gettext, :ex_aws, :hackney, :poison]]
  end

  defp deps do
    [{:phoenix, "~> 1.2.1"},
     {:phoenix_pubsub, "~> 1.0"},
     {:phoenix_html, "~> 2.6"},
     {:phoenix_live_reload, "~> 1.0", only: :dev},
     {:gettext, "~> 0.11"},
     {:cowboy, "~> 1.0"},
     {:ex_aws, "~> 1.0.0-beta0"},
     {:poison, "~> 2.0"},
     {:hackney, "~> 1.6"}]
  end
`

mix.lock
'  "ex_aws": {:hex, :ex_aws, "1.0.0-beta1", "7c101ce714190b91cc7033c4adc3eb593a5daea17cb4444e33e18d824dd0ee7f", [:mix], [{:httpoison, "~> 0.8", [hex: :httpoison, optional: true]}, {:jsx, "~> 2.5", [hex: :jsx, optional: true]}, {:poison, "~> 1.2 or ~> 2.0", [hex: :poison, optional: true]}, {:sweet_xml, "~> 0.5", [hex: :sweet_xml, optional: true]}]},
'

**Comment by benwilson512** [comment]
How are you starting your iex session? Are you doing `iex -S mix` ?

**Comment by schummjm** [comment]
Ahhh.. I was just doing `iex`.  Rookie mistake...

Now I'm getting this error with Hackney:

```
** (exit) exited in: GenServer.call(ExAws.Config.AuthCache, {:refresh_config, %{access_key_id: [{:system, "XXXXX"}, :instance_role], host: {"$region", "dynamodb.$region.amazonaws.com"}, http_client: ExAws.Request.Hackney, json_codec: Poison, port: 80, region: "us-west-2", scheme: "https://", secret_access_key: [{:system, "XXXXX"}, :instance_role]}}, 5000)
    ** (EXIT) an exception was raised:
        ** (UndefinedFunctionError) undefined function ExAws.Request.Hackney.request/2 (module ExAws.Request.Hackney is not available)
            ExAws.Request.Hackney.request(:get, "http://169.254.169.254/latest/meta-data//iam/security-credentials/")

```

Is there something off in my config?

```
config :ex_aws,
  access_key_id: [{:system, "REMOVED"}, :instance_role],
  secret_access_key: [{:system, "REMOVED"}, :instance_role],
  http_client: ExAws.Request.Hackney,
  json_codec: Poison,
  region: "us-west-2",
  port: 443
```

**Comment by benwilson512** [comment]
Ah you should be on `1.0.0-beta1` not `beta0`, I believe that's a bug from there.

**Comment by schummjm** [comment]
I really appreciate your prompt help :)

I changed to `1.0.0-beta1` and still have the same error.

mix.exs
```
  defp deps do
    [{:phoenix, "~> 1.2.1"},
     {:phoenix_pubsub, "~> 1.0"},
     {:phoenix_html, "~> 2.6"},
     {:phoenix_live_reload, "~> 1.0", only: :dev},
     {:gettext, "~> 0.11"},
     {:cowboy, "~> 1.0"},
     {:ex_aws, "~> 1.0.0-beta1"},
     {:hackney, "~> 1.6"},
     {:poison, "~> 2.0"},]
  end
```

mix.lock
`  "ex_aws": {:hex, :ex_aws, "1.0.0-beta1", "7c101ce714190b91cc7033c4adc3eb593a5daea17cb4444e33e18d824dd0ee7f", [:mix], [{:httpoison, "~> 0.8", [hex: :httpoison, optional: true]}, {:jsx, "~> 2.5", [hex: :jsx, optional: true]}, {:poison, "~> 1.2 or ~> 2.0", [hex: :poison, optional: true]}, {:sweet_xml, "~> 0.5", [hex: :sweet_xml, optional: true]}]},`

**Comment by benwilson512** [comment]
Ah oops apparently I fixed that in master but never released. If you update to `beta3` you should hopefully be good to go!

The issue is that right around when the beta came out I switched default HTTP providers but missed one of the default configuration settings.

**Comment by schummjm** [comment]
Thanks for the update!  I had to update Elixir as well to make it work with the update to gen_stage.

Hoping you can point me in the right direction on what's happening on this new issue.. I feel like I'm so close...

I tried running this a couple ways, but having issues with the `|> ExAws.request!`.  This very well could just be my lack of Elixir knowledge.

```
Interactive Elixir (1.3.2) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)> alias ExAws.Dynamo
ExAws.Dynamo
iex(2)>  Dynamo.create_table("Users", "email", %{email: :string}, 1, 1)
%ExAws.Operation.JSON{before_request: nil,
 data: %{"AttributeDefinitions" => [%{"AttributeName" => :email,
      "AttributeType" => "S"}],
   "KeySchema" => [%{"AttributeName" => "email", "KeyType" => "HASH"}],
   "ProvisionedThroughput" => %{"ReadCapacityUnits" => 1,
     "WriteCapacityUnits" => 1}, "TableName" => "Users"},
 headers: [{"x-amz-target", "DynamoDB_20120810.CreateTable"},
  {"content-type", "application/x-amz-json-1.0"}], http_method: :post,
 parser: #Function<1.10035553/1 in ExAws.Operation.JSON.new/2>, path: "/",
 service: :dynamodb, stream_builder: nil}
iex(3)>   |> ExAws.request!
** (SyntaxError) iex:3: syntax error before: '|>'
    
iex(3)> Dynamo.create_table("Users", "email", %{email: :string}, 1, 1) |> ExAws.request!
** (exit) exited in: GenServer.call(ExAws.Config.AuthCache, {:refresh_config, %{access_key_id: [{:system, "REMOVED"}, :instance_role], debug_requests: true, host: "localhost", http_client: ExAws.Request.Hackney, json_codec: Poison, port: 8000, region: "us-west-2", scheme: "http://", secret_access_key: [{:system, "REMOVED"}, :instance_role]}}, 5000)
    ** (EXIT) time out
    (elixir) lib/gen_server.ex:604: GenServer.call/3
    (ex_aws) lib/ex_aws/config.ex:62: ExAws.Config.retrieve_runtime_value/2
    (elixir) lib/stream.ex:490: anonymous fn/4 in Stream.map/2
    (elixir) lib/enum.ex:2843: Enumerable.List.reduce/3
    (elixir) lib/stream.ex:1284: Enumerable.Stream.do_each/4
    (elixir) lib/enum.ex:839: Enum.find/3
    (ex_aws) lib/ex_aws/config.ex:50: anonymous fn/2 in ExAws.Config.retrieve_runtime_config/1
    (stdlib) lists.erl:1263: :lists.foldl/3
    (ex_aws) lib/ex_aws/config.ex:30: ExAws.Config.new/2
    (ex_aws) lib/ex_aws.ex:29: ExAws.request/2
    (ex_aws) lib/ex_aws.ex:41: ExAws.request!/2
iex(3)> [error] GenServer ExAws.Config.AuthCache terminating
** (CaseClauseError) no case clause matching: {:error, %{reason: :connect_timeout}}
    (ex_aws) lib/ex_aws/instance_meta.ex:12: ExAws.InstanceMeta.request/2
    (ex_aws) lib/ex_aws/instance_meta.ex:22: ExAws.InstanceMeta.security_credentials/1
    (ex_aws) lib/ex_aws/config/auth_cache.ex:37: ExAws.Config.AuthCache.refresh_config/2
    (ex_aws) lib/ex_aws/config/auth_cache.ex:27: ExAws.Config.AuthCache.handle_call/3
    (stdlib) gen_server.erl:615: :gen_server.try_handle_call/4
    (stdlib) gen_server.erl:647: :gen_server.handle_msg/5
    (stdlib) proc_lib.erl:247: :proc_lib.init_p_do_apply/3
Last message: {:refresh_config, %{access_key_id: [{:system, "REMOVED"}, :instance_role], debug_requests: true, host: "localhost", http_client: ExAws.Request.Hackney, json_codec: Poison, port: 8000, region: "us-west-2", scheme: "http://", secret_access_key: [{:system, "REMOVED"}, :instance_role]}}
State: ExAws.Config.AuthCache
```

**Comment by benwilson512** [comment]
TL;DR: You aren't configuring your AWS keys properly.

You have it configured right now to do `[{:system, "REMOVED"}, :instance_role]`. What this means is that it will do `System.get_env("REMOVED")` to retrieve the data from your system environment. If that returns nil, it'll try to grab the instance role meta-data. That however only works on AWS EC2 instances.

Long story short, whatever you actually have instead of `"REMOVED"`, do `System.get_env("REMOVED")` and if that returns nil, you haven't actually set an environment variable named that.
